### ISSUE TITLE
Tutorial

### ISSUE THREAD

**Issue by Albatros3** [dridri/bcflight#2]
Hi and a big hello from france.
Great job you're done.
Have you got a tutorial or a link for schemas and installations please?
Thank you.

**Comment by dridri** [comment]
I just wrote the beginning of the first draft : https://github.com/dridri/bcflight/wiki/Building
I hope it will help and will continue it as soon as possible.
Tell me if you have any questions :D

**Comment by smarkoco** [comment]
Amazing work!! Would love to get this drone setup integrated with my planned "Control a Drone Over the Internet" concept for a web application (eventually on cloudchop.net).

Are there perhaps any updated tutorials or list of dependencies/requirements for building? Would also be great if there were a couple brief explanations of the overall information flows, for example, I was not sure about how the control signals are sent, guessing the LoRa Radio? Do you think it's possible to receive a buffer of frames on a desktop and forward them to a remote browser, and also receive flight commands from that same browser over a bidirectional socket (e.g. with [Tornado](https://www.tornadoweb.org/en/stable/)) or would the closed-loop latencies be too excrutiating?

Was also wondering what drone build you used or would recommend?
Thanks so much!

**Comment by dridri** [comment]
Hello, thanks for your interest in this project !

As you could see on the latest commits, I'm getting my hands on this project after a very long break and now working to make it more reliable. There's indeed a lot of work on documentation needed, this should grow in the next weeks.

The current readme is unclear about radio, you can actually use any kind of link you want : it's currently possible to use some FSK/LoRa modules, UDP/IP over regular wifi, and raw wifi using compatible dongles (it sends raw packets without need to associate with ground station). It's also possible to handle other radio communication systems by inheriting the [Link](https://github.com/dridri/bcflight/blob/master/libcontroller/links/Link.h#L71) class and overloading the few virtual functions it exposes.

The video feed can be sent back via legacy analog VTX with ~50ms glass-to-glass latency, but also via raw-wifi / regular-wifi with an extra cost of ~30ms and ~50ms latency respectively.

When I started this project I only used the wifi/raw-wifi to send and receive both control signals and video feedback, the code is dusty and needs a massive rewrite but what you are asking for is totally possible with 200~250ms total latency which should be just Ok if you intend to fly it smoothly.
The overall system could be something like : client browser → websocket → server & ground-station → wifi/raw-wifi → drone
The [libcontroller](https://github.com/dridri/bcflight/blob/master/libcontroller) is intended to be used in this kind of project, it has just the minimum required code to communicate with the drone from ground and should be easy to include in any program.

My current setup uses the carrier board described in the [Hardware](../#hardware) section with some 20$ ESCs and motors, a TrampHV VTX directly connected to Raspberry Pi's composite video output and the newest Raspberry Pi Camera Module 3, nothing more.

TL;DR : what you are asking for is possible but not in a reliable way for the moment as the code is old and faulty, but I hope to progress quickly during the next weeks

**Comment by smarkoco** [comment]
Thanks so much for the in depth reply, and happy Sunday! no worries at all about the documentation, I can see you've already put in a ton of work, and when a project is still in flux it can just be too time-consuming to constantly update documentation lol

Luckily I was able to get the `controller_pc` module to build and run last night on my Ubuntu Studio 22.04 system! Just had to install a few packages with `apt` when the `make` command threw some helpful "module not found" errors (they were usually related to *-dev packages for wireless, etc.). 

For some reason, the below commands from the files [HERE](https://github.com/dridri/bcflight/search?q=%3Ccodec_api.h%3E) were not able to find that package even though I eventually explicitly added the `bcflight/external` path to my include path (??)

```
extern "C" {
#include <codec_api.h>
}
```

so I had to replace with the below to get it working 
```
extern "C" {
#include "../../external/openh264-master/codec/api/wels/codec_api.h"
}
```
**Dependency installs:**
I also just ended up building and installing any other missing packages from github repos.

1. **Qt**: 
    - `sudo apt-get install qtmultimedia5-dev install qtbase5-dev qtchooser qt5-qmake qtbase5-dev-tools`
2. **nasm**:
    - `sudo apt install nasm`
3. **QScintilla**: Download source from [HERE](https://riverbankcomputing.com/software/qscintilla/download). Then run:
    - `tar -xzf  QScintilla_src-2.13.4.tar.gz`
    - `cd QScintilla_src-2.13.4/src`
    - `qmake`
    - `make`
    - `make install`
4. **MP4V2**: repo located [HERE](https://github.com/enzo1982/mp4v2)
    - `git clone https://github.com/enzo1982/mp4v2.git`
    - `cd mp4v2`
    - `cmake . && make`
    - `make install`
5. **Shine**: It was a little tricky for me to find the right "shine" repo [HERE](https://github.com/toots/shine)
    - `git clone https://github.com/toots/shine`
    - `cd shine`
    - `autoreconf --install --force`
    - `automake`
    - `./configure`
    - `make`
    - `make install`

After these installations, your `Find*.cmake` scripts in `bcflight/controller_pc/CMakeModules` worked perfectly :+1::
    - `git clone https://github.com/dridri/bcflight.git`
    - `cd bcflight/controller_pc`
    - `cmake -DCMAKE_BUILD_TYPE=Release -S <git clone path here>/bcflight/controller_pc -B <git clone path here>/bcflight/controller_pc/build`
    - `cd build`
    - `make -j4`
    - `./controller_pc`

:tada: 
![image](https://user-images.githubusercontent.com/53878603/221422138-01d33f46-fb8c-4d7f-9b44-c717a14a2745.png)


**Next Steps:**
- I just ordered the camera and compute board, and will try getting the video feed working first.  
- Sadly the BMP180 and [BMP280](https://www.mouser.com/datasheet/2/783/BST_BMP280_DS001-1509562.pdf) are now "obsolete", and none of their [other models](https://www.digikey.com/en/products/filter/pressure-sensors-transducers/512?s=N4IgTCBcDaIEIFkAKACJAlAogZWwVSxW0wDlsB5dEAXQF8g) have the 8SMD package, so I'll probably delay getting that piece. Seems like they upgraded to either Bosch [BMP390](https://www.mouser.com/datasheet/2/783/bst_bmp390_ds002-2448819.pdf) or [BMP581](https://www.mouser.com/pdfDocs/Bosch_BMP581_DS.pdf)
- ordering [ICM-42605](https://www.mouser.com/ProductDetail/410-ICM-42605)
- I'll probably order your board from OSHPark.com (looks like what you did from the purple board :grin: )
- What would you recommend as a ground station? Or can my wifi interface just communicate directly with the [RFM95W](https://www.mouser.com/ProductDetail/RF-Solutions/RFM95W-915S2?qs=sGAEpiMZZMu3sxpa5v1qriuzrjO9DfPiQuQApVEeBDo%3D) for initial testing (sorry kinda new to the drone world)
- Finally I'm having a little trouble figuring out which pins are the "video composite out", would you mind letting me know? Are these to be accessed on the your pcb board or the cm4 itself?

Thanks a bunch, will let you know how it all goes!

**Comment by smarkoco** [comment]
Yeah for sure! that would be great to see it added to the Readme.md/Wiki, feel free to let me know if I should Pull Request it (then you could edit)

Would love to see `ffmpeg` get used, I love that package, I use it in my work in motion tracking/frame processing/video conversion, and a lot of the commands are highly parallelized

Thanks for pointing out the RPi camera connector, was actually searching eBay for something for that haha :sweat_smile:, yes I was actually about to ask about all the essential connectors earlier, I think I may try to prototype a bit with the Regular RPi4B if I need to access more peripherals (USB/etc), luckily I do have one of those laying around the house

I'll definitely give the regular TCP/UDP route a try first, and will keep those other adapters in mind for the future. Thanks for all the tips!

Would love to start contributing if I can, I do have an xbox controller, and can definitely give that a try! I was also thinking of trying to map some keyboard input to flight control, such as how a [helicopter is flown in Call of Duty](https://www.reddit.com/r/joinsquad/comments/d9ijw6/helicopter_controls_arent_great_heres_my/), but that might be a bad idea for reasons I can't think of

**Comment by dridri** [comment]
Commit 25299a8d3 brings back video live stream feature from drone to PC controller app. I'll add a description into the wiki, but in the meantime if you're curious you can take a look at the config.lua file [here](../blob/master/flight/config.lua#L173) and [there](../blob/master/flight/config.lua#L200).

With 8mbps bitrate there seem to be less than 60ms glass-to-glass latency over ethernet, libcamera+v4l2 is surprisingly good compared to old closed-source drivers and OpenMAX (was 120+ over wifi at only 2mbps)
